src/TokenHandler/AccountManager.mo

```diff
@@ -0,0 +1,408 @@
+
+      return ?depositDelta;
+    };
+
+    /// Processes the consolidation transfer for a principal.
+    func processConsolidationTransfer(p : Principal, deposit : Nat) : async* {
+      #Ok : Nat;
+      #Err : ICRC1.TransferError or {
+        #CallIcrc1LedgerError;
+        #InsufficientDeposit;
+      };
+    } {
+      if (deposit <= fee_) return #Err(#InsufficientDeposit);
+
+      let transferAmount : Nat = Int.abs(deposit - fee_);
+
+      let transferResult = try {
+        await icrc1Ledger.icrc1_transfer({
+          from_subaccount = ?Mapping.toSubaccount(p);
+          to = { owner = ownPrincipal; subaccount = null };
+          amount = transferAmount;
+          fee = ?fee_;
+          memo = null;
+          created_at_time = null;
+        });
+      } catch (_) {
+        #Err(#CallIcrc1LedgerError);
+      };
+
+      switch (transferResult) {
+        case (#Ok _) {
+          ignore updateDeposit(p, 0);
+          totalConsolidated_ += transferAmount;
+          journal.push((Time.now(), p, #consolidated({ deducted = deposit; credited = transferAmount })));
+        };
+        case (#Err err) {
+          journal.push((Time.now(), p, #consolidationError(err)));
+        };
+      };
+
+      transferResult;
+    };
+
+    /// Attempts to consolidate the funds for a particular principal.
+    func consolidate(p : Principal) : async* () {
+      let latestDeposit = try { await* loadDeposit(p) } catch (err) {
+        pushToBacklog(p, 0);
+        return;
+      };
+
+      ignore updateDeposit(p, latestDeposit);
+
+      ignore depositRegistry.set(
+        p,
+        func(info) {
+          info.underway := latestDeposit;
+          underwayFunds += latestDeposit - info.underway;
+          true;
+        },
+      );
+
+      let transferResult = await* processConsolidationTransfer(p, latestDeposit);
+
+      switch (transferResult) {
+        case (#Err(#InsufficientDeposit)) {
+          ignore creditRegistry.debit(p, latestDeposit - fee_);
+        };
+        case (#Err(#BadFee { expected_fee })) {
+
+          ignore creditRegistry.debit(p, latestDeposit - fee_);
+          ignore creditRegistry.credit(p, latestDeposit - expected_fee);
+
+          fee_ := expected_fee;
+
+          journal.push((Time.now(), ownPrincipal, #feeUpdated({ old = fee_; new = expected_fee })));
+
+          recalculateBacklog();
+
+          let retryResult = await* processConsolidationTransfer(p, latestDeposit);
+          switch (retryResult) {
+            case (#Err(#InsufficientDeposit)) {
+              ignore creditRegistry.debit(p, latestDeposit - fee_);
+            };
+            case (#Err _) {
+              ignore creditRegistry.debit(p, latestDeposit - fee_);
+              pushToBacklog(p, latestDeposit);
+            };
+            case (_) {};
+          };
+        };
+        case (#Err _) {
+          ignore creditRegistry.debit(p, latestDeposit - fee_);
```

Comment:
Like above, I don't think there should be a debit because the consolidation failed.